[REVIEW] Enable gather/slicing/joins with ListColumns in Python

shwina opened the pull request:
Depends on https://github.com/rapidsai/cudf/pull/5816

- [x] Adds a `cudf::column(cudf::column_view)` constructor for lists columns (`copying.cu`)
- [x] Adds tests for gather/slice/join with LIstColumns in Python

This actually started off as https://github.com/rapidsai/cudf/pull/5785 and https://github.com/rapidsai/cudf/pull/5820, but I split those off into smaller PRs because they were needed by #5723.

Fixes https://github.com/rapidsai/cudf/issues/5621

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

shwina commented on cpp/src/lists/copying/slice.cu:
```diff
@@ -0,0 +1,26 @@
+                                    cudaStream_t stream,
+                                    rmm::mr::device_memory_resource* mr)
+{
+  auto sliced_table = cudf::detail::gather(table_view{{list.parent()}},
```
Just a temporary implementation

shwina replied:
fixed

shwina commented:
Just remembered I have a libcudf TODO here :)

shwina commented on python/cudf/cudf/core/column/column.py:
```diff
@@ -880,6 +880,11 @@ def distinct_count(self, method="sort", dropna=True):
+            if not is_list_dtype(self.dtype):
+                raise ValueError(f"Cannot cast list column to {dtype}")
+            else:
+                return self
```
Yeah good catch. Is that basically the only kind of casting we should can/support for list columns, i.e., casting the leaf values?

kkraus14 replied:
Only other semi-normal thing I can think of would be being able to do `String` <--> `List(int8)` including at various levels of nesting.

Otherwise in a bit more niche area, you could imagine someone wanting to `view` as opposed to `cast` from say `int32` --> `int64` assuming there's a divisible number of elements in each List.

shwina commented on cpp/include/cudf/lists/detail/slice.hpp:
```diff
@@ -0,0 +1,30 @@
+namespace cudf {
+namespace lists {
+namespace detail {
+std::unique_ptr<cudf::column> slice(lists_column_view const& list,
```
Where should tests for this go?

harrism replied:
tests/lists/slice_tests.cpp ???

rgsl888prabhu replied:
`cpp/tests/copying`

shwina replied:
Since this is really an internal function that is only ever used by the `column(column_view)` constructor, I put tests in `column_tests.cu`. I'm not sure about the name `slice` - since `slice` is also used for the API that does zero-copy slicing - but that's the name that the equivalent strings function uses. @davidwendt would you have any advice here?

davidwendt replied:
The name of the strings function I think you are referring to in libcudf is actually called [`cudf::strings::slice_strings()`](https://docs.rapids.ai/api/libcudf/stable/group__strings__substring.html#gaa689719250048ea788b5db1549c8265f) so as to not confuse it with `cudf::slice()`

[There is a `cudf::strings::detail::slice()` that is called by `column(column_view)`](https://github.com/rapidsai/cudf/blob/ab1a1e8964f5e5b30f8aa08e81764234f24024cf/cpp/src/column/column.cu#L187) -- in the same way as `cudf::lists::detail::slice()` in [this PR](https://github.com/rapidsai/cudf/blob/ab1a1e8964f5e5b30f8aa08e81764234f24024cf/cpp/src/column/column.cu#L240).

I think adding the tests for lists for `column(column_view)` into `column_test.cu` as you have done makes sense to me.

shwina replied:
Right -- I guess my worry is that the `detail::slice()` functions could be confused with the public `cudf::slice()` function, and I was wondering whether we should change both `strings::detail::slice()` and `lists::detail::slice()` to something else (`copy_slice()` perhaps?)

davidwendt replied:
I think renaming these to `detail::copy_slice()` is a good idea. The `cudf::strings::detail::slice()` was just created before `cudf::slice()` was created/ported.

nvdbaranec replied:
Thoroughly agree.

harrism commented on cpp/include/cudf/lists/detail/gather.cuh:
```diff
@@ -112,6 +107,11 @@ gather_data make_gather_data(cudf::lists_column_view const& source_column,
                       return src_offsets[index];
                     });
 
+  // now that we are doing using the gather_map, we can release the underlying prev_base_offsets.
```
Do you mean "done" rather than "doing"?

shwina replied:
This change is from PR #5816 - sorry for the confusion :( cc @nvdbaranec

harrism commented on cpp/include/cudf/lists/detail/slice.hpp:
```diff
@@ -0,0 +1,30 @@
+namespace cudf {
+namespace lists {
+namespace detail {
+std::unique_ptr<cudf::column> slice(lists_column_view const& list,
```
```suggestion
std::unique_ptr<cudf::column> slice(lists_column_view const& lists,
```

harrism commented on cpp/src/lists/copying/slice.cu:
```diff
@@ -0,0 +1,68 @@
+                                    cudaStream_t stream,
+                                    rmm::mr::device_memory_resource* mr)
+{
+  size_type lists_count = lists.size();
+  if (lists_count == 0) { return cudf::empty_like(lists.parent()); }
+  if (end < 0 || end > lists_count) end = lists_count;
+  CUDF_EXPECTS(((start >= 0) && (start < end)), "Invalid start parameter value.");
+  lists_count = end - start;
```
Avoid initialize-then-modify (ITM)

```suggestion
  if (lists.is_empty() == 0) { return cudf::empty_like(lists.parent()); }
  if (end < 0 || end > lists.size()) end = lists.size();
  CUDF_EXPECTS(((start >= 0) && (start < end)), "Invalid slice range.");
  auto lists_count = end - start;
```

harrism commented on cpp/src/lists/copying/slice.cu:
```diff
@@ -0,0 +1,68 @@
+{
+  size_type lists_count = lists.size();
+  if (lists_count == 0) { return cudf::empty_like(lists.parent()); }
+  if (end < 0 || end > lists_count) end = lists_count;
```
Do we need to be this flexible? Treat all negative values of `end` the same?

shwina replied:
Borrowed pattern from strings [`slice`](https://github.com/rapidsai/cudf/blob/branch-0.15/cpp/src/strings/copying/copying.cu#L31). Happy to throw here instead.

harrism commented on cpp/src/lists/copying/slice.cu:
```diff
@@ -0,0 +1,68 @@
+  CUDF_EXPECTS(((start >= 0) && (start < end)), "Invalid start parameter value.");
+  lists_count = end - start;
+
+  auto execpol = rmm::exec_policy(stream);
```
Define just before use.

shwina replied:
Done

harrism commented on cpp/src/lists/copying/slice.cu:
```diff
@@ -0,0 +1,68 @@
+    &end_offset, offsets_data + end, sizeof(cudf::size_type), cudaMemcpyDeviceToHost, stream));
+
+  // compute the offsets column of the result:
+  rmm::device_uvector<cudf::size_type> out_offsets(lists_count + 1, stream);
```
Why the +1? (comment).

Seems you only use lists_count+1, never just lists_count, so maybe put it into the definition of lists_count and comment there.

shwina replied:
Fixed. I introduced `offsets_count` since we do actually use just `lists_count` below.

harrism commented on cpp/src/lists/copying/slice.cu:
```diff
@@ -0,0 +1,68 @@
+
+  thrust::transform(execpol->on(stream),
+                    offsets_data + start,
+                    offsets_data + end + 1,
```
Why the +1 here? // comment.

shwina replied:
Done

harrism commented on cpp/src/lists/copying/slice.cu:
```diff
@@ -0,0 +1,68 @@
+    cudf::data_type{cudf::type_id::INT32}, lists_count + 1, out_offsets.release());
+
+  // compute the child column of the result:
+  auto child =
```
This is a little hard to follow. Maybe a comment to explain what the two different versions of slice() being called are.

shwina replied:
Added a comment

harrism commented on cpp/tests/column/column_test.cu:
```diff
@@ -403,4 +403,53 @@ TYPED_TEST(TypedColumnTest, ColumnViewConstructorWithMask)
   EXPECT_NE(original_view.null_mask(), copy_view.null_mask());
 }
 
```
Test the range expectations, negative start, etc.

shwina replied:
I can't test that as it's a detail of the `lists::detail::copy_slice` function (`lists::detail::slice`) function, that is not exposed in any public API.

harrism replied:
I see that array_tests.[cpp, cu] have explicit calls like this `auto results      = cudf::strings::detail::copy_slice(strings_view, start, end);`.  So clearly the tests are testing `strings::detail::copy_slice`. So why can't we test `lists:detail::copy_slice`?

shwina replied:
Because the `slice` function is an implementation detail, and [testing behaviour, not implementation](
https://testing.googleblog.com/2013/08/testing-on-toilet-test-behavior-not.html?m=1) is considered good practice.

harrism replied:
So by "can't", you mean "don't think I should"?

harrism replied:
I guess I'm confused about why there needs to be a *behavior* that `CUDF_EXPECTS` something, but we are not allowed to test that behavior.  If we can expect a behavior, we should be able to test it.

shwina replied:
> So by "can't", you mean "don't think I should"?

Yes -- apologies, I should have been more clear and explained my reasoning better. 

> I guess I'm confused about why there needs to be a behavior that CUDF_EXPECTS something, but we are not allowed to test that behavior. If we can expect a behavior, we should be able to test it.

TL;DR: I think we should test that behaviour if (and only if) it is exposed via a public API.

Long answer:

At the moment, we have a "private" `detail::copy_slice` function, only a part of whose functionality is exercised by a "public" `cudf::column` constructor. We have tests for that part. It's true that now we have untested code in libcudf, but it's OK because that code can not be exercised in any way using the libcudf API. If/when it is exposed publicly (e.g., in a `cudf::copy_slice`API), _then_ it becomes necessary to test for that behaviour.

In my opinion, testing that behaviour while it is private is actively harmful; consider the following scenario:

Suppose we wrote a test for the range checking the behaviour of `detail::copy_slice` today. Suppose that tomorrow, we made a public `cudf::copy_slice`. Another test for the range checking behaviour of `cudf::copy_slice` would be redundant, and understandably, could be skipped. At that point, we would have a dangerous situation where public behaviour is tested in a test for a private function, the latter of which could be dropped or changed at any time.

harrism replied:
OK, thanks for clarifying. I understand now.

nvdbaranec commented on cpp/src/lists/copying/copying.cu:
```diff
@@ -0,0 +1,71 @@
+  cudf::size_type start_offset{0};
+  cudf::size_type end_offset{0};
+  auto offsets_data = lists.offsets().data<cudf::size_type>();
+  CUDA_TRY(cudaMemcpyAsync(
```
Use get_value (https://github.com/rapidsai/cudf/blob/b33383c1fa644f07feb2d96aa16caf4e0445fa56/cpp/include/cudf/detail/get_value.cuh#L41) here.
```suggestion
start_offset = cudf::detail::get_value<size_type>(lists.offsets(), start, stream)
```

nvdbaranec commented on cpp/src/lists/copying/copying.cu:
```diff
@@ -0,0 +1,71 @@
+  auto offsets_data = lists.offsets().data<cudf::size_type>();
+  CUDA_TRY(cudaMemcpyAsync(
+    &start_offset, offsets_data + start, sizeof(cudf::size_type), cudaMemcpyDeviceToHost, stream));
+  CUDA_TRY(cudaMemcpyAsync(
```
```suggestion
end_offset = cudf::detail::get_value<size_type>(lists.offsets(), end, stream)
```

nvdbaranec commented on cpp/src/lists/copying/copying.cu:
```diff
@@ -0,0 +1,71 @@
+  auto child =
+    (lists.child().type() == cudf::data_type{type_id::LIST})
+      ? copy_slice(lists_column_view(lists.child()), start_offset, end_offset, stream, mr)
+      : std::make_unique<cudf::column>(
```
I haven't posted the PR yet, but I have a change in place that adds a function to lists_column_view called get_sliced_child(), which will return a column_view that has the root column offset applied.  It has some smarts in it to do the minimum amount of retrieving of offsets off the gpu.  Just worth being advised of.

shwina replied:
I'd like to get this PR merged in ASAP if that's OK. Happy to update this to use `get_sliced_child()` in a follow-up, since I currently have a cascade of dependent PRs and I don't want to add yet another dependency

nvdbaranec replied:
Yes, absolutely.  Just a reference for the future (I think the fate of get_sliced_child() vs. child() is probably still up for debate in that PR).

shwina commented:
rerun tests

shwina commented:
@harrism I'm going to retarget this to branch-0.15 which will require a force-push. Before I do that, could you please re-review/approve, as I don't want your review comments to be lost

shwina commented on cpp/src/lists/copying/copying.cu:
```diff
@@ -0,0 +1,67 @@
+          cudf::detail::slice(lists.child(), {start_offset, end_offset}, stream).front());
+
+  // Compute the null mask of the result:
+  auto null_mask = cudf::copy_bitmask(lists.null_mask(), start_offset, end_offset, stream, mr);
```
@rgsl888prabhu caught that here we should probably be doing instead:

```
  auto null_mask = cudf::copy_bitmask(lists.null_mask(), start, end, stream, mr);
```

harrism commented:
> @harrism I'm going to retarget this to branch-0.15 which will require a force-push. Before I do that, could you please re-review/approve, as I don't want your review comments to be lost?

See my most recent comment about testing before I approve.

harrism commented:
Seems to already be on branch-0.15. Why is a rebase needed? If you mean branch-0.16, a merge should be sufficient.

shwina commented:
> Seems to already be on branch-0.15. Why is a rebase needed? If you mean branch-0.16, a merge should be sufficient.

Initially, this PR had base branch 0.15. Then, I changed it to 0.16. After that, I changed it again back to 0.15. As it stands, this PR wants to merge commits from 0.16 into 0.15 right now, but after I force-push my local (rebased) branch, it won't do that.

kkraus14 commented:
@shwina can you comment here once you've rebased so we make sure that's done before we merge? 😄

rgsl888prabhu commented:
It still needs a fix for mask copy which @shwina mentioned https://github.com/rapidsai/cudf/pull/5771#discussion_r468864588

shwina commented:
@kkraus14 @rgsl888prabhu Force pushed, and also addressed the mask copying issue

rgsl888prabhu commented:
#5816 has been merged, may be we can remove the dependency label.
